carapace-server/src/main/java/org/carapaceproxy/api/CertificatesResource.java

```diff
@@ -165,82 +172,54 @@ private CertificateBean findCertificateById(String certId) {
             }
             return certBean;
         }
 
         return null;
     }
 
-    static String stateToStatusString(DynamicCertificateState state) {
-        if (state == null) {
-            return "unknown";
-        }
-        switch (state) {
-            case WAITING:
-                return "waiting"; // certificate waiting for issuing/renews
-            case VERIFYING:
-                return "verifying"; // challenge verification by LE pending
-            case VERIFIED:
-                return "verified"; // challenge succeded
-            case ORDERING:
-                return "ordering"; // certificate order pending
-            case REQUEST_FAILED:
-                return "request failed"; // challenge/order failed
-            case AVAILABLE:
-                return "available";// certificate available(saved) and not expired
-            case EXPIRED:     // certificate expired
-                return "expired";
-            default:
-                return "unknown";
-        }
-    }
-
-    static String stateToStatusString(CertificateMode mode) {
-        if (mode == null) {
-            return "unknown";
-        }
-        switch (mode) {
-            case STATIC:
-                return "static";
-            case ACME:
-                return "acme";
-            case MANUAL:
-                return "manual";
-            default:
-                return "unknown";
-        }
-    }
-
     @POST
     @Path("{domain}/upload")
     @Consumes(MediaType.APPLICATION_OCTET_STREAM)
-    public Response uploadCertificate(@PathParam("domain") String domain, InputStream uploadedInputStream) throws Exception {
+    public Response uploadCertificate(
+            @PathParam("domain") String domain,
+            @QueryParam("type") @DefaultValue("manual") String type,
+            InputStream uploadedInputStream) throws Exception {
 
+        // Certificate type (manual | acme)
+        CertificateMode certType = stringToCertificateMode(type);
+        if (certType == null || STATIC.equals(certType)) {
+            return Response.status(422).entity("ERROR: illegal type of certificate. Available: manual, acme").build();
+        }
+        // Certificate content (optional)
         byte[] data = uploadedInputStream.readAllBytes();
-
-        // Validation
-        if (!CertificatesUtils.validateKeystore(data)) {
+        if (data != null && data.length > 0 && !CertificatesUtils.validateKeystore(data)) {
             return Response.status(422).entity("ERROR: unable to read uploded certificate.").build();
         }
 
-        CertificateData cert = new CertificateData(
-                domain, "", Base64.getEncoder().encodeToString(data), AVAILABLE.name(), "", "", true
-        );
-        cert.setManual(true);
-
+        String encodedData = "";
+        DynamicCertificateState state = WAITING;
+        boolean available = false;
+        if (data != null && data.length > 0) {
```

Comment:
is it possible to have a MANUAL certificate without the contents of the certificate ?